### ISSUE TITLE
[HOW-TO] Install on Kodi 19-Matrix

### ISSUE THREAD

**Issue by sy6sy2** [Catch-up-TV-and-More/plugin.video.catchuptvandmore#363]
Currently Catch-Up TV & More is not officially available on Kodi 19 because some others add-ons that we use as dependancies are not yet available on the official Matrix repository.

Once all dependancies that we need will be available, we will submit the plugin to the official Kodi repository like we currently do for Kodi 17 and Kodi 18.

If you really need to use Kodi 19 alpha version, this is the steps to follow in order to be able to use Catch-Up TV & More (at your own risk):

1. Download Youtube-DL Kodi 19 version zip file from this [URL](https://github.com/AnonTester/script.module.youtube.dl/archive/master-kodi19.zip)
2. Open Kodi, go to `Add-ons --> Install from zip file` and select the Youtube-DL zip file (if needed, allow "Unknown sources")
3. Close Kodi
3. Download Codequick Kodi 19 version from this [URL](https://github.com/willforde/script.module.codequick/archive/v19.zip) and unzip it.
4. Open the freshly unzipped folder and move `script.module.codequick` folder in your Kodi `addons` folder (see https://kodi.wiki/view/Kodi_data_folder for the location)
5. From this [directory](https://github.com/Catch-up-TV-and-More/repository.catchuptvandmore/tree/master/zips/krypton_beta/plugin.video.catchuptvandmore) download the latest Catch-Up TV & More zip file, unzip it and apply these patches:
    * In `addon.xml` file replace `<import addon="xbmc.python" version="2.25.0"/>` by `<import addon="xbmc.python" version="3.0.0"/>`
6. Move `plugin.video.catchuptvandmore` folder in your Kodi `addons` folder
7. Open Kodi
8. Go to `Settings --> System`, move from `Standard` to `Expert` view. In the current setting window go to `Add-ons --> Manage dependencies`
9. Open `CodeQuick --> Dependencies --> HTMLement` and choose `Install`
10. Enable CodeQuick
11. Go to `Main menu --> Add-ons --> My add-ons --> Videos add-ons --> Catch-Up TV & More --> Dependecnies` and install all not installed dependencies
12. Enable Catch-Up TV & More

**Note that Catch-Up TV & More will not update by itself. To update it you have to reapply steps 6 and 7 manually**

_Tested 16th November 2020 with Kodi 19 Alpha 3_

**Comment by enigma131** [comment]
Concerning point 2 you can't install the zip as is.
You have to unzip it, then replace in lib/yd_private_libs/util.py the instances xbmc.LOGNOTICE by xbmc.LOGDEBUG . After Move script.module.youtube.dl folder in your Kodi addons folder.
Without this some channel including no DRM ones don't play and gives an error.

**Comment by sy6sy2** [comment]
Thank you for the detail.
You can also install the zip, and only after the install is done you can find and replace LOGNOTICE instances 😉 
Both order are possible, I will update the HOW-TO.

**Comment by enigma131** [comment]
I've compile Kodi 19 alpha 2 and the plugin is still working :+1:

**Comment by enigma131** [comment]
I have upgraded my Fedora to version 33 that came with V19 Alpha1 natively.
The procedure didn't work, other stuff is missing ....
Quasi all dependency's won't load, example : htmlement: python 2.25.0 could not be satisfied

**Comment by enigma131** [comment]
I have solved the dependencys by copying addon folder from Centos8 installation.

Now all plugings load, but still python problem: 
script.module.youtube.dl/lib:/usr/lib64/python39.zip:/usr/lib64/python3.9:/usr/lib64/python3.9/lib-dynload:/usr/lib64/python3.9/site-packages:/usr/lib/python3.9/site-packages
2020-11-01 10:38:15.200 T:4315    DEBUG <general>: CPythonInvoker(0, ~/.kodi/addons/plugin.video.catchuptvandmore/service.py): entering source directory /~/.kodi/addons/plugin.video.catchuptvandmore
2020-11-01 10:38:15.200 T:4315    DEBUG <general>: CPythonInvoker(0, ~/.kodi/addons/plugin.video.catchuptvandmore/service.py): instantiating addon using automatically obtained id of "plugin.video.catchuptvandmore" dependent on version 3.0.0 of the xbmc.python api

**Comment by enigma131** [comment]
I've reinstalled Kodi on Fedora33.
After the tutorial above, Catchup tv was playing.
But there is a but .... I have filled my accounts, and after click on m6, i got error.
(It is working on centos and kodi 19)
error now on each pluging load : 
.kodi/addons/service.xbmc.versioncheck/resources/lib/runner.py): instantiating addon using automatically obtained id of "service.xbmc.versioncheck" dependent on version 3.0.0 of the xbmc.python api
2020-11-01 14:42:45.334 T:14826   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'ModuleNotFoundError'>
                                                   Error Contents: No module named 'kodi_six'
                                                   Traceback (most recent call last):
                                                     File "/home/gilles/.kodi/addons/plugin.video.catchuptvandmore/service.py", line 23, in <module>
                                                       from kodi_six import xbmcaddon
                                                   ModuleNotFoundError: No module named 'kodi_six'
                                                   -->End of Python script error report<--

**Comment by enigma131** [comment]
Still no Idea ?

**Comment by sy6sy2** [comment]
Sorry didn't have time.
Kodi-six seems to be missing, try install this one if you are on Matrix https://mirrors.kodi.tv/addons/matrix/script.module.kodi-six/script.module.kodi-six-0.1.3.1.zip

**Comment by enigma131** [comment]
Hi, plugin updated.
2 new missings: 
ModuleNotFoundError: No module named 'tzlocal'
ModuleNotFoundError: No module named 'pyqrcode'
found and updated from same repo

Now i have other errors : 
ModuleNotFoundError: No module named 'inputstreamhelper'

and : 

raise RouteMissing("unable to import route module: %s" % module_path)
codequick.support.RouteMissing: 'unable to import route module: resources.lib.channels.fr.francetv'

**Comment by sy6sy2** [comment]
The one from your system package manager. RPM on Fedora, right ?
The name is `kodi-inputstream-adaptive`

**Comment by enigma131** [comment]
yep:
Paquets disponibles
Nom          : kodi-inputstream-adaptive
Version      : 2.6.0
Publication  : 1.fc33
Architecture : x86_64
Taille       : 740 k
Source       : kodi-inputstream-adaptive-2.6.0-1.fc33.src.rpm
Dépôt        : rpmfusion-free
Résumé       : Adaptive file addon for Kodi's InputStream interface
URL          : https://github.com/peak3d/inputstream.adaptive/
Licence      : GPLv2+ and BSD
Description  : Adaptive file addon for Kodi's InputStream interface.

Installed but not up to date, did not decrypt

**Comment by enigma131** [comment]
I've try to recompile from source with branch Matrix, got an error:

[ 88%] Linking CXX static library libmpegts.a
[ 88%] Built target mpegts
Scanning dependencies of target inputstream.adaptive
[ 88%] Building CXX object CMakeFiles/inputstream.adaptive.dir/src/main.cpp.o
~/kodi_project/inputstream.adaptive/src/main.cpp:44:10: erreur fatale: kodi/addon-instance/inputstream/StreamCodec.h : Aucun fichier ou dossier de ce type
   44 | #include <kodi/addon-instance/inputstream/StreamCodec.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminée.

**Comment by enigma131** [comment]
So no way to handle with RPMfusion nor compiled one

**Comment by sy6sy2** [comment]
If you want to use Kodi 19 you need all binary addons (kodi-inputstream-adaptive, kodi.binary.instance.inputstream, ...) to be compatible with Kodi 19. It means you need to use Matrix version from Jenkins (https://jenkins.kodi.tv).

Why do you no use Kodi 18?

**Comment by enigma131** [comment]
As you know I still have python problems on my Fedora 32 with kodi 18. On this disk installation I got invitation to upgrade to Fedora 33. I did not on this disk, but I have cloned this partition to a fresh disk and accept the upgrade. As I write above, Fedora 33 comes natively with Kodi 19, I haven't forced anything ...
yum info kodi
Dernière vérification de l’expiration des métadonnées effectuée il y a 0:00:16 le lun. 09 nov. 2020 08:23:04.
Paquets installés
Nom          : kodi
Version      : 19.0
Publication  : 0.5.20201005git54be31b.fc33
Architecture : x86_64
Taille       : 0.0  
Source       : kodi-19.0-0.5.20201005git54be31b.fc33.src.rpm
Dépôt        : @System
Depuis le dé : rpmfusion-free-updates
Résumé       : Media center
URL          : http://www.kodi.tv/
Licence      : GPLv2+ and GPLv3+ and LGPLv2+ and BSD and MIT
Description  : Kodi is a free cross-platform media-player jukebox and entertainment hub.
             : Kodi can play a spectrum of of multimedia formats, and featuring playlist,
             : audio visualizations, slideshow, and weather forecast functions, together
             : third-party plugins.
             : 
             : This is a meta package.
After installation of Inputstream from Fedora 33 repo, in the log above I have 404 IP error, why ? Is something useful in my log ?

Concerning my compilation for inputstream , I've taken the source from: 
**git clone --branch Matrix https://github.com/peak3d/inputstream.adaptive.git**
Is it false ?

I don't find a source for linux here : 
https://jenkins.kodi.tv/view/Addons/job/peak3d/job/inputstream.adaptive/job/Matrix/lastSuccessfulBuild/

Complicated Fedora with Kodi's plugins...

**Comment by enigma131** [comment]
I suppose I must wait Kodi 19 final to have support for Fedora ?

**Comment by sy6sy2** [comment]
You use the right branch for ISA.

BTW, it seems that Kodi 18.9 is available on Fedora 33: https://fedora.pkgs.org/33/unitedrpms-x86_64/kodi-18.9-7.fc33.x86_64.rpm.html

**Comment by enigma131** [comment]
Thanks for the link!
As Kodi 18 and Kodi 19 can't coexist on same Fedora 33 distribution, I've decided to test 18.9 on Fedora 33 fresh VM install and it works (low resolution for some channels of course). It's a third party compilation, so I don't want to uninstall kodi 19 from this disk, I will test it on spare disk with Fedora 32 , upgrade first Fedora to  33 and install this RPM on real material from my PC (can't access to my DVB cards from VM) ... I tell you informed, it's out of topic from that thread that is dedicated to Kodi 19. Keep in mind Kodi 19 Beta 1 will be out soon....

**Comment by enigma131** [comment]
There is also a kodi-inputstream-adaptive-2.4.5-7.fc33.x86_64.rpm on this site. I have done a try but:
DEBUG <general>: ------ Window Deinit (DialogConfirm.xml) ------
2020-11-11 18:32:39.658 T:5628    ERROR <general>: Add-on 'inputstream.adaptive' is not compatible with Kodi
2020-11-11 18:32:39.675 T:5519    DEBUG <general>: ------ Window Init (DialogNotification.xml) ------
2020-11-11 18:32:39.717 T:5628    ERROR <general>: EXCEPTION: Unknown addon id 'inputstream.adaptive'.
2020-11-11 18:32:39.717 T:5628    DEBUG <general>: [script.module.inputstreamhelper] inputstream.adaptive None is installed and enabled.
2020-11-11 18:32:39.717 T:5628    ERROR <general>: EXCEPTION: Unknown addon id 'inputstream.adaptive'.

**Comment by enigma131** [comment]
I know now why inputstream.adaptive 2.6.0.1 from Fedora 33 repo don't work. I have downloaded the corresponded sources and they are from may the 19th .And Inputstream was corrected en of july for M6 problem ... So must wait for an upgrade

**Comment by enigma131** [comment]
I know now why inputstream.adaptive 2.6.0.1 from Fedora 33 repo don't work. I have downloaded the corresponded sources and they are from may the 19th .And Inputstream was corrected en of july for M6 problem ... So must wait for an upgrade

**Comment by enigma131** [comment]
I got Kodi Alpha 3 this morning from Fedora updates. I have removed inputstream.adaptive installed from repo (still NOT updated) .
I extract source from official repo kodi-19.0-0.6.20201031gitbb0699d.fc33.src.rpm
Recompile inputstream.adaptive from git, install it from zip and now it WORKS !
Now i have compatibility with plugin pvr.hts but this is a other story...

**Comment by enigma131** [comment]
pvr.hts recompiled. I have now Catchup TV and more with inputstream.adaptive, pvr.hts and Youtube working on Fedora 33 alpha 3. For your information beta1 reported to november the 14th

**Comment by wwark** [comment]
Good to hear ! Needs a lot of effort to survive in Fedora world.
One time I tried Arch Linux but after two times updates my workstation is not working well so I stick now on debian/ubuntu.

**Comment by enigma131** [comment]
Yes Arch is an other story. If you have DVB cards build via DKMS like me, keep in mind that this distribution don't update previous working kernel on updates. Very dangerous, one kernel update a week on rolling releases. Fedora bakups the kernel. I try Fedora to learn deeply Rehat system, my goal is to implement CentOs on my DVB server (10 years support)

**Comment by berkhornet** [comment]
Hi. I've installed the matrix version on Kodi Matrix Alpha 3as per instructions above. When trying to use Quest OD I get a JSON error on all menu options. Relevant extract from log is

2020-11-16 11:48:49.560 T:612     DEBUG <general>: CInputManager::HandleKey: return (0xf00d) pressed, action is Select
2020-11-16 11:48:49.562 T:612     DEBUG <general>: CGUIMediaWindow::GetDirectory (plugin://plugin.video.catchuptvandmore/resources/lib/channels/uk/questod/list_programs_mode/?_pickle_=80049556000000000000007d94288c075f7469746c655f948c0c4c454156494e4720534f4f4e948c076974656d5f6964948c0771756573746f64948c1363617465676f72795f6e616d655f76616c7565948c0b6c6173742d6368616e636594752e)
2020-11-16 11:48:49.562 T:612     DEBUG <general>:   ParentPath = [plugin://plugin.video.catchuptvandmore/resources/lib/channels/uk/questod/list_categories/?_pickle_=8004952e000000000000007d94288c075f7469746c655f948c085175657374204f44948c076974656d5f6964948c0771756573746f6494752e]
2020-11-16 11:48:49.563 T:2156    DEBUG <general>: Thread waiting start, auto delete: false
2020-11-16 11:48:49.568 T:13184   DEBUG <general>: CAddonDatabase::SetLastUsed[plugin.video.catchuptvandmore] took 6 ms
2020-11-16 11:48:49.569 T:2156    DEBUG <general>: XFILE::CPluginDirectory::StartScript - calling plugin Catch-up TV & More('plugin://plugin.video.catchuptvandmore/resources/lib/channels/uk/questod/list_programs_mode/','5','?_pickle_=80049556000000000000007d94288c075f7469746c655f948c0c4c454156494e4720534f4f4e948c076974656d5f6964948c0771756573746f64948c1363617465676f72795f6e616d655f76616c7565948c0b6c6173742d6368616e636594752e','resume:false')
2020-11-16 11:48:49.569 T:924     DEBUG <general>: Thread LanguageInvoker start, auto delete: false
2020-11-16 11:48:49.570 T:924      INFO <general>: initializing python engine.
2020-11-16 11:48:49.570 T:924     DEBUG <general>: CPythonInvoker(69, C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\addon.py): start processing
2020-11-16 11:48:49.649 T:924     DEBUG <general>: -->Python Interpreter Initialized<--
2020-11-16 11:48:49.649 T:924     DEBUG <general>: 
                                                   
2020-11-16 11:48:49.649 T:924     DEBUG <general>: CPythonInvoker(69, C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\addon.py): the source file to load is "C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\addon.py"
2020-11-16 11:48:49.649 T:924     DEBUG <general>: CPythonInvoker(69, C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\addon.py): setting the Python path to C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.addon.signals\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.certifi\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.chardet\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.future\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.htmlement\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.idna\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.inputstreamhelper\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.kodi-six\libs;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.pyqrcode\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.pytz\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.requests\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.six\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.tzlocal\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.urllib3\lib;C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.youtube.dl\lib;C:\Program Files\Kodi\system\python\Lib;C:\Program Files\Kodi\python3.8.zip;C:\Program Files\Kodi\system\python\DLLs;C:\Program Files\Kodi;C:\Program Files\Kodi\system\python;C:\Program Files\Kodi\system\python\lib\site-packages
2020-11-16 11:48:49.649 T:924     DEBUG <general>: CPythonInvoker(69, C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\addon.py): entering source directory C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore
2020-11-16 11:48:49.649 T:924     DEBUG <general>: CPythonInvoker(69, C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\addon.py): instantiating addon using automatically obtained id of "plugin.video.catchuptvandmore" dependent on version 3.0.0 of the xbmc.python api
2020-11-16 11:48:49.664 T:612     DEBUG <general>: ------ Window Init (DialogBusy.xml) ------
2020-11-16 11:48:50.095 T:924   WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-11-16 11:48:50.096 T:924     DEBUG <general>: [Catch-up-TV-&-More.support] Dispatching to route: '/resources/lib/channels/uk/questod/list_programs_mode/'
2020-11-16 11:48:50.096 T:924     DEBUG <general>: [Catch-up-TV-&-More.support] Callback parameters: '{'item_id': 'questod', 'category_name_value': 'last-chance'}'
2020-11-16 11:48:50.096 T:924     DEBUG <general>: [Catch-up-TV-&-More.support] Attempting to import route: resources.lib.channels.uk.questod
2020-11-16 11:48:50.303 T:924   WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-11-16 11:48:50.303 T:924     DEBUG <general>: [urlquick] Cache location: C:\Users\mickr\AppData\Roaming\Kodi\userdata\addon_data\plugin.video.catchuptvandmore\
2020-11-16 11:48:50.312 T:924   WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-11-16 11:48:50.313 T:924   WARNING <general>: xbmc.translatePath is deprecated and might be removed in future kodi versions. Please use xbmcvfs.translatePath instead.
2020-11-16 11:48:50.467 T:924     DEBUG <general>: [urlquick] Requesting resource: https://www.questod.co.uk/api/shows/last-chance
2020-11-16 11:48:50.468 T:924     DEBUG <general>: [urlquick] Request headers: {'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate', 'Accept-language': 'en-gb,en-us,en', 'Connection': 'keep-alive', 'Host': 'www.questod.co.uk'}
2020-11-16 11:48:50.852 T:924     DEBUG <general>: [urlquick] Caching 301 Moved Permanently response
2020-11-16 11:48:50.860 T:924     DEBUG <general>: [urlquick] Redirecting to = https://www.dplay.co.uk/api/shows/last-chance
2020-11-16 11:48:51.198 T:924     DEBUG <general>: [urlquick] Caching 301 Moved Permanently response
2020-11-16 11:48:51.200 T:924     DEBUG <general>: [urlquick] Redirecting to = https://www.discoveryplus.co.uk/api/shows/last-chance
2020-11-16 11:48:51.402 T:924     DEBUG <general>: [urlquick] Caching 200 OK response
2020-11-16 11:48:51.412 T:924     DEBUG <general>: [urlquick] Attempting to decode with default encoding: 'utf8'
2020-11-16 11:48:51.415 T:924     ERROR <general>: [Catch-up-TV-&-More.support] Expecting value: line 1 column 1 (char 0)
                                                   Traceback (most recent call last):
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib\codequick\support.py", line 328, in run_callback
                                                       redirect = parent_ins._process_results(results)
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib\codequick\route.py", line 95, in _process_results
                                                       raw_listitems = validate_listitems(raw_listitems)
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib\codequick\route.py", line 34, in validate_listitems
                                                       raw_listitems = list(raw_listitems)
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\resources\lib\channels\uk\questod.py", line 118, in list_programs_mode
                                                       json_parser = json.loads(resp.text)
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\__init__.py", line 357, in loads
                                                       return _default_decoder.decode(s)
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 337, in decode
                                                       obj, end = self.raw_decode(s, idx=_w(s, 0).end())
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 355, in raw_decode
                                                       raise JSONDecodeError("Expecting value", s, err.value) from None
                                                   json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

Many thanks for your help

**Comment by wwark** [comment]
Hi @berkhornet,
QuestOD is not used anymore on the last version of the plugin.
We switched to dplay.
What is the version of the plugin used ?

**Comment by sy6sy2** [comment]
My fault @wwark,
The CUTV version pointed by the steps above are no up to date.
I will updates the steps today to include this change and also the new v19 Codequick version 😉

**Comment by berkhornet** [comment]
Many thanks for your prompt response. I will check the v19 instructions later today and try again.

Regards Mick

________________________________

**Comment by sy6sy2** [comment]
Steps above are now up to date and work with Kodi 19 alpha 3.
I also just committed a fix in order to have adaptive/DRM streams working on Kodi 19 (https://github.com/Catch-up-TV-and-More/plugin.video.catchuptvandmore/commit/70ef0e32e9edce9a888a36f4cae29c934e616e59)

**Comment by berkhornet** [comment]
Hi. Many thanks for the updated versions. I installed CUTV 0.2.28 beta 4 and Code Quick 1.0 as per instructions. I now have the *DPlay option. But I'm still getting the JSON decode error with any of the options within Dplay. 

2020-11-16 14:31:43.699 T:16020   ERROR <general>: [Catch-up-TV-&-More.support] Expecting value: line 1 column 1 (char 0)
                                                   Traceback (most recent call last):
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib\codequick\support.py", line 328, in run_callback
                                                       redirect = parent_ins._process_results(results)
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib\codequick\route.py", line 95, in _process_results
                                                       raw_listitems = validate_listitems(raw_listitems)
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\script.module.codequick\lib\codequick\route.py", line 34, in validate_listitems
                                                       raw_listitems = list(raw_listitems)
                                                     File "C:\Users\mickr\AppData\Roaming\Kodi\addons\plugin.video.catchuptvandmore\resources\lib\channels\uk\dplay.py", line 132, in list_programs_mode
                                                       json_parser = json.loads(resp.text)
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\__init__.py", line 357, in loads
                                                       return _default_decoder.decode(s)
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 337, in decode
                                                       obj, end = self.raw_decode(s, idx=_w(s, 0).end())
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 355, in raw_decode
                                                       raise JSONDecodeError("Expecting value", s, err.value) from None
                                                   json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

Let me know if you need any more information. Thanks.

**Comment by enigma131** [comment]
I've got finally good (fresh recompiled) versions for kodi-pvr-hts and kodi-inputstream-adaptive from rpmfusion (Fedora 33), all is still working

**Comment by 0x0040** [comment]
System like @enigma131 (Fedora 33, with Kodi 19.0 beta 1). I recompile the inputstream-streamadaptive to v2.6.6 and it works fine until i start "France TV" and got the error : 'HTMLParser' object has no attribute 'unescape'

Indeed with python 3.9 the unescape() method in the html.parser.HTMLParser class has been removed (it was deprecated since Python 3.4). html.unescape() should be used for converting character references to the corresponding unicode characters. (source: https://docs.python.org/3/whatsnew/3.9.html#removed)

**Comment by enigma131** [comment]
Yes I haven't recompile IS, it was working fine for kodi alpha3, but after they have put out beta3 and the Fedora 33 team didn't check the corresponding plugins that don't wok any more. see the mess here for pvr.hts same problem: https://bugzilla.rpmfusion.org/show_bug.cgi?id=5836

**Comment by wwark** [comment]
Thank you @0x0040 and @enigma131 
We used HTMLParser just for France TV.
I will opened a ticket dedicated to this issue.

**Comment by enigma131** [comment]
France TV corrected with update from today. Great job !

**Comment by enigma131** [comment]
Got update to kodi 19 beta 2 . All is still working.

**Comment by enigma131** [comment]
Arte seems need correction as kodi 18 branch

**Comment by enigma131** [comment]
Kodi 19 Rc1 is out, final version for february the 14 th:  https://github.com/xbmc/xbmc/milestone/140

**Comment by enigma131** [comment]
Need correction too for France TV, I don't watch TF1 but I just see is out too on Kodi 19 Rc1 and Fedora 33

**Comment by jeff2900** [comment]
Hello,
Suggestion for you
If kodi v19 compatibility is conditioned by python version why do not change following lines in **addon.xml** file.
Replace: 
```
  <requires>
    <import addon="xbmc.python" version="2.25.0"/>
  ...
```
By:
```
  <requires>
    <trycatch>
       <try>
          <import addon="xbmc.python" version="3.0.0"/>
       </try>
       <catch>
          <import addon="xbmc.python" version="2.25.0"/>
       </catch>
    </trycatch>
   ...
```

**Comment by wwark** [comment]
Hi @jeff2900,

The main issue to propose CUTV&More Kodi 19 in the official repo Kodi 19 or own repo is some other plugin's (YT-DL, ...) are not present in the official repo Kodi 19. 
We proposed YT-DL to official repo Kodi 19 and we waiting this plugin to be approved for some weeks ...

Thanks,

**Comment by dekani** [comment]
Hello,

It seems they are requesting new changes for YT-DL before validating it. (Check pull request  last status)
https://github.com/xbmc/repo-scripts/pull/1836

Best regards.

**Comment by enigma131** [comment]
I've updated the plugin, modified again version for python 3.0.0 in addon.xml, and voila all channels available again ! thanks

**Comment by sy6sy2** [comment]
Hi all,
Sorry I was a bit busy since some days.
I will try to apply YT-dl changes ASAP so that Willforde will be able to submit CodeQuick on the repo and finally we will be able to officially submit CUTV&M on the repo 😉 

Sorry for the delay and sorry to not be ready in the same time as Kodi 19.

**Comment by sy6sy2** [comment]
Hi all, 
So to summarize the actual situation:
* script.module.youtube.dl-21.204.0+matrix.1 is now available on Kodi official repo for all Matrix users (https://mirrors.kodi.tv/addons/matrix/script.module.youtube.dl/)
* script.module.codequick-1.0.1+matrix.1 s now available on Kodi official repo for all Matrix users (https://mirrors.kodi.tv/addons/matrix/script.module.codequick/)
* Our last beta release (0.2.32~beta01) should be compatible (I mean installable) on Kodi 17, Kodi 18 and 19 because all dependency add-ons/scripts are available on Kodi official repo

It means that all users who are using our Krypton beta channel repository (see https://catch-up-tv-and-more.github.io/installation/#ii-add-the-official-catch-up-tv--more-repository) should be able to install CUTV&M on Matrix without any "hack" or "tricks".

The add-on will be available soon for all Matrix users on our official stable repository and on the official Kodi repository.

Thank you @wwark for your hard work on the plugin, thank you @willforde for the CodeQuick framework and thank you to all users who are using the add-ons and reports issues!
Enjoy your week-end 😉

**Comment by wwark** [comment]
Great News !

**Comment by enigma131** [comment]
Fine.
I've installed via this method on fresh installed Linux Mint 20.1 (Ubuntu 2004 LTS based)
All is working  except M6 DRM channels. Widewine is installed, (was requred by Tf1).
Here part of log : 
2021-02-27 11:32:02.408 T:4142    ERROR <general>: [Catch-up-TV-&-More.support] sequence item 1: expected a bytes-like object, str found
                                                   Traceback (most recent call last):
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/codequick/support.py", line 306, in run_callback
                                                       redirect = parent_ins(route, arg_params, self.callback_params)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/codequick/resolver.py", line 67, in __call__
                                                       results = super(Resolver, self).__call__(route, args, kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/codequick/script.py", line 169, in __call__
                                                       return route.function(self, *args, **kwargs)
                                                     File "/home/toto/.kodi/addons/plugin.video.catchuptvandmore/resources/lib/channels/fr/6play.py", line 550, in get_live_url
                                                       resp2 = urlquick.post(
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/urlquick.py", line 505, in post
                                                       return request('post', url, data=data, json=json, **kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/urlquick.py", line 482, in request
                                                       return s.request(method=method, url=url, **kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/urlquick.py", line 436, in request
                                                       response = super(Session, self).request(*args, **kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.requests/lib/requests/sessions.py", line 533, in request
                                                       resp = self.send(prep, **send_kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/urlquick.py", line 445, in send
                                                       return super(Session, self).send(request, **kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.requests/lib/requests/sessions.py", line 646, in send
                                                       r = adapter.send(request, **kwargs)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/urlquick.py", line 349, in send
                                                       urlhash = hash_url(request)
                                                     File "/home/toto/.kodi/addons/script.module.codequick/lib/urlquick.py", line 203, in hash_url
                                                       return hashlib.sha1(b''.join((data, req.body or b''))).hexdigest()
                                                   TypeError: sequence item 1: expected a bytes-like object, str found

**Comment by wwark** [comment]
Hi @enigma131,
This thread is dedicated to install Kodi 19 Matrix.
Could you please opened a new ticket ?

Thanks,

**Comment by enigma131** [comment]
It's concerning Kodi 19 Matrix install. Why a new ticket . A new ticket only for drm ?

**Comment by wwark** [comment]
Yes the issue is only on Kodi 19.
It is not an issue to install Kodi 19 so I think a new will be better. We will close tomorrow this ticket.
The issue is not for DRM but using session requests we opened a ticket to codequick (the framework we used).

**Comment by enigma131** [comment]
Ok well understood, I will do it :+1:

**Comment by romain211289** [comment]
Bonjour,

Merci pour le travail ! Je viens de passer à K19 au vu de votre annonce. Après plusieurs essais : Replay OK sans DRM/compte utilisateur , Replay OK NRJ12 /

**REPLAY 6PLAY : KO (aucuns lancements)**

Pour le direct TV : 

- Erreur lors de l'obtention du guide
- **Quasiment toutes les chaines fonctionnent sauf celle du groupe M6**

J'ai vérifié à plusieurs reprises si cela ne venait pas de mon compte 6Play, mais je vous confirme que non car sur KODI 18 aucuns soucis.

- Vidéo du net : OK

Bonne journée !

**Comment by wwark** [comment]
Bonjour @romain211289,

Le soucis du groupe M6 est suivi sur ce ticket https://github.com/Catch-up-TV-and-More/plugin.video.catchuptvandmore/issues/447

I close this ticket because this method of installing the plugin of Kodi 19 is not required any more. You can opened you ticket to describe your issue.
Je ferme ce ticket parce que cette méthode d'installation du plugin dans Kodi 19 n'est plus requise maintenant. Vous pouvez ouvrir un nouveau ticket si vous rencontrez des soucis en décrivant le problème.

Merci,
Wwark
